pkg/rootless/rootless_linux.c

```diff
@@ -0,0 +1,846 @@
+static const char *_unprivileged_user_namespaces = "/proc/sys/kernel/unprivileged_userns_clone";
+
+static int open_files_max_fd;
+fd_set open_files_set;
+static uid_t rootless_uid_init;
+static gid_t rootless_gid_init;
+
+static int
+syscall_setresuid (uid_t ruid, uid_t euid, uid_t suid)
+{
+  return (int) syscall (__NR_setresuid, ruid, euid, suid);
+}
+
+static int
+syscall_setresgid (gid_t rgid, gid_t egid, gid_t sgid)
+{
+  return (int) syscall (__NR_setresgid, rgid, egid, sgid);
+}
+
+uid_t
+rootless_uid ()
+{
+  return rootless_uid_init;
+}
+
+uid_t
+rootless_gid ()
+{
+  return rootless_gid_init;
+}
+
+static void
+do_pause ()
+{
+  int i;
+  struct sigaction act;
+  int const sig[] =
+    {
+     SIGALRM, SIGHUP, SIGINT, SIGPIPE, SIGQUIT, SIGPOLL,
+     SIGPROF, SIGVTALRM, SIGXCPU, SIGXFSZ, 0
+    };
+
+  act.sa_handler = SIG_IGN;
+
+  for (i = 0; sig[i]; i++)
+    sigaction (sig[i], &act, NULL);
+
+  prctl (PR_SET_NAME, "podman pause", NULL, NULL, NULL);
+  while (1)
+    pause ();
+}
+
+static char **
+get_cmd_line_args (pid_t pid)
+{
+  int fd;
+  char path[PATH_MAX];
+  char *buffer;
+  size_t allocated;
+  size_t used = 0;
+  int ret;
+  int i, argc = 0;
+  char **argv;
+
+  if (pid)
+    sprintf (path, "/proc/%d/cmdline", pid);
+  else
+    strcpy (path, "/proc/self/cmdline");
+  fd = open (path, O_RDONLY);
+  if (fd < 0)
+    return NULL;
+
+  allocated = 512;
+  buffer = malloc (allocated);
+  if (buffer == NULL)
+    return NULL;
+  for (;;)
+    {
+      ret = TEMP_FAILURE_RETRY (read (fd, buffer + used, allocated - used));
+      if (ret < 0)
+        {
+          free (buffer);
+          return NULL;
+        }
+
+      if (ret == 0)
+        break;
+
+      used += ret;
+      if (allocated == used)
+        {
+          allocated += 512;
+          char *tmp = realloc (buffer, allocated);
+          if (tmp == NULL)
+            {
+              free (buffer);
+              return NULL;
+            }
+	  buffer = tmp;
+        }
+    }
+  close (fd);
+
+  for (i = 0; i < used; i++)
+    if (buffer[i] == '\0')
+      argc++;
+  if (argc == 0)
+    {
+      free (buffer);
+      return NULL;
+    }
+
+  argv = malloc (sizeof (char *) * (argc + 1));
+  if (argv == NULL)
+    {
+      free (buffer);
+      return NULL;
+    }
+  argc = 0;
+
+  argv[argc++] = buffer;
+  for (i = 0; i < used - 1; i++)
+    if (buffer[i] == '\0')
+      argv[argc++] = buffer + i + 1;
+
+  argv[argc] = NULL;
+
+  return argv;
+}
+
+static bool
+can_use_shortcut ()
+{
+  int argc;
+  char **argv;
+  bool ret = true;
+
+#ifdef DISABLE_JOIN_SHORTCUT
+  return false;
+#endif
+
+  argv = get_cmd_line_args (0);
+  if (argv == NULL)
+    return false;
+
+  for (argc = 0; argv[argc]; argc++)
+    {
+      if (argc == 0 || argv[argc][0] == '-')
+        continue;
+
+      if (strcmp (argv[argc], "mount") == 0
+          || strcmp (argv[argc], "search") == 0
+          || strcmp (argv[argc], "system") == 0)
```

Comment:
I don’t know about the rest of the reexec logic, but at least this just can’t be used in a generic library package.